src/Kernel.php

```diff
@@ -21,6 +27,11 @@ public static function getInstance(): Kernel
         return static::$instance;
     }
 
+    public function run(): void
```

Comment:
Przeczytaj dopiero po przeczytaniu komentarza który oznaczyłem na początku #exception!!

Skoro stworzyłeś już swój wyjątek i wyrzucasz go musisz go teraz przechwycić w tym miejscu.
służy do tego tak zwany try catch.

owiń wywołanie $this->router->route();
w try catch i przechwyć wyjątek UndefinedControllerException następnie zloguj to loggerem i wypisz na ekranie wiadomość błędu za pomocą echo $e->getMessage();